jetty-util/src/main/java/org/eclipse/jetty/util/compression/CompressionPool.java

```diff
@@ -56,73 +65,102 @@ public CompressionPool(int capacity)
     /**
      * @return Object taken from the pool if it is not empty or a newly created Object
      */
-    public T acquire()
+    public Entry acquire()
     {
-        T object;
-
-        if (_capacity == 0)
-            object = newObject();
-        else
+        Entry entry = null;
+        if (_pool != null)
         {
-            object = _pool.poll();
-            if (object == null)
-                object = newObject();
-            else if (_capacity > 0)
-                _numObjects.decrementAndGet();
+            Pool<Entry>.Entry acquiredEntry = _pool.acquire(e -> new Entry(newObject()));
+            if (acquiredEntry != null)
+            {
+                entry = acquiredEntry.getPooled();
+                entry.setEntry(acquiredEntry);
+            }
         }
 
-        return object;
+        return (entry == null) ? new Entry(newObject()) : entry;
     }
 
     /**
-     * @param object returns this Object to the pool or calls {@link #end(Object)} if the pool is full.
+     * @param entry returns this Object to the pool or calls {@link #end(Object)} if the pool is full.
      */
-    public void release(T object)
+    public void release(Entry entry)
     {
-        if (object == null)
-            return;
+        entry.release();
+    }
 
-        if (_capacity == 0 || !isRunning())
+    @Override
+    protected void doStart() throws Exception
+    {
+        if (_capacity > 0)
+            _pool = new Pool<>(Pool.StrategyType.RANDOM, _capacity, true);
+        super.doStart();
+    }
+
+    @Override
+    public void doStop() throws Exception
+    {
+        _pool.close();
+        _pool = null;
+        super.doStop();
+    }
+
+    public class Entry implements Closeable
+    {
+        private T _value;
+        private Pool<Entry>.Entry _entry;
```

Comment:
make value and entry final

Reply:
Shouldn't this be non-final so we can null it out when closed.

Reply:
No need to null out, the whole instance will be discarded soon after close. 
The close impls should be robust with multiple closes ... actually it probably NPEs currently, so final is better anyway.